File: WatchBox/Shared/Models/MediaDetail.swift
//
//  MediaDetail.swift
//  SceneBox
//
//  Created by SpontaneousArray on 30.07.26.
//

import Foundation

nonisolated public struct MediaDetail: Identifiable, Sendable {
    public let id: String
    public let type: MediaType
    public let name: String
    public let year: String?
    public let posterURL: URL?
    public let backdropURL: URL?
    public let logoURL: URL?
    public let description: String?
    public let cast: [String]
    public let directors: [String]
    public let writers: [String]
    public let genres: [String]
    public let imdbRating: Double?
    public let runtime: String?
    public let released: Date?
    public let country: String?
    public let awards: String?
    public let episodes: [Episode]
    public var moviedbID: Int?
    public var trailerYouTubeIDs: [String] = []

    public var trailerURL: URL? {
        trailerYouTubeIDs.first.flatMap { URL(string: "https://www.youtube.com/watch?v=\($0)") }
    }

    public var seasons: [Int] {
        let numbered = Set(episodes.map(\.season))
        return numbered.filter { $0 > 0 }.sorted() + (numbered.contains(0) ? [0] : [])
    }

    public func episodes(inSeason season: Int) -> [Episode] {
        episodes.filter { $0.season == season }.sorted { $0.episode < $1.episode }
    }

    public var summaryLine: String {
        [year, runtime, genres.prefix(3).joined(separator: ", ")]
            .compactMap { $0 }
            .filter { !$0.isEmpty }
            .joined(separator: "  ·  ")
    }
}

File: lib/core/models/media_detail.dart
import 'episode.dart';
import 'media_type.dart';

class MediaDetail {
  MediaDetail({
    required this.id,
    required this.type,
    required this.name,
    this.year,
    this.posterUrl,
    this.backdropUrl,
    this.logoUrl,
    this.description,
    this.cast = const [],
    this.directors = const [],
    this.writers = const [],
    this.genres = const [],
    this.imdbRating,
    this.runtime,
    this.released,
    this.country,
    this.awards,
    this.episodes = const [],
    this.moviedbId,
    this.trailerYouTubeIds = const [],
  });

  final String id;
  final MediaType type;
  final String name;
  final String? year;
  final Uri? posterUrl;
  final Uri? backdropUrl;
  final Uri? logoUrl;
  final String? description;
  final List<String> cast;
  final List<String> directors;
  final List<String> writers;
  final List<String> genres;
  final double? imdbRating;
  final String? runtime;
  final DateTime? released;
  final String? country;
  final String? awards;
  final List<Episode> episodes;
  final int? moviedbId;
  final List<String> trailerYouTubeIds;

  Uri? get trailerUrl => trailerYouTubeIds.isEmpty
      ? null
      : Uri.parse('https://www.youtube.com/watch?v=${trailerYouTubeIds.first}');

  List<int> get seasons => {
        for (final episode in episodes)
          if (episode.season > 0) episode.season,
      }.toList()
        ..sort();

  List<Episode> episodesInSeason(int season) => episodes
      .where((episode) => episode.season == season)
      .toList()
    ..sort((a, b) => a.episode.compareTo(b.episode));

  String get summaryLine => [
        year,
        runtime,
        if (genres.isNotEmpty) genres.take(3).join(', '),
      ].whereType<String>().where((value) => value.isNotEmpty).join('  ·  ');
}
